### PR TITLE
Fix ZeroDivisionError with "* * R/0 * *"

### PR DESCRIPTION
**Pull request by cuu508**
In relation to https://github.com/taichino/croniter/issues/164

Checks for zero divisor and raises CroniterBadCronError

### CODE REVIEW

File: src/croniter/croniter.py [hunk 1]
```diff
@@ -851,7 +851,12 @@ def do(self, idx, hash_type="h", hash_id=None, range_end=None, range_begin=None)
             crc = random.randint(0, 0xFFFFFFFF)
         else:
             crc = binascii.crc32(hash_id) & 0xFFFFFFFF
-        return ((crc >> idx) % (range_end - range_begin + 1)) + range_begin
```
Comment (kiorky) [inline src/croniter/croniter.py:854]: LGTM,
Small detail: can you include range boundaries in the log message for better user interaction ?

Comment (cuu508) [reply src/croniter/croniter.py:854]: Improved the exception message a bit, it now looks like so:

    croniter.croniter.CroniterBadCronError: Bad expression: r/0

It's not perfect - it doesn't preserve the r / R and h / H case.

File: src/croniter/croniter.py [hunk 2]
```diff
@@ -841,25 +841,27 @@ class HashExpander:
...
+        if rng == 0:
+            raise CroniterBadCronError("Bad expression: %s" % expr)
+
+        return ((crc >> idx) % rng) + range_begin
 
     def expand(self, efl, idx, expr, hash_id=None, match='', **kw):
         """Expand a hashed/random expression to its normal representation"""
         if match == '':
-            match = self.match(efl, idx, expr, hash_id, **kw)
+            match = hash_expression_re.match(expr)
```
Comment (kiorky) [inline src/croniter/croniter.py:864]: Oh sorry , i read at first too quickly the PR !!!


The proposed changes break the factory/utility pattern, `match` should not trigger any exception, it only matchs if the expression appears to be  consumed by the plugin, and it's the case here.

The one which should trigger the error when called is `hashexpander.expand`.
`hashexpander.expand` should call `match` to get the RE groupdict elements and process forward, either by raising an exception if the expression is considered invalid, or then process forward.

Comment (cuu508) [reply src/croniter/croniter.py]: I reverted the change to `HashExpander.match()`, and moved the range check to `HashExpander.expand()`.

File: src/croniter/croniter.py [hunk 3]
```diff
@@ -892,6 +895,9 @@ def expand(self, efl, idx, expr, hash_id=None, match='', **kw):
             )
         elif m['divisor']:
             # Example: H/15 -> 7-59/15 (i.e. 7,22,37,52)
+            if int(m["divisor"]) == 0:
+                raise CroniterBadCronError("Bad expression: %s" % expr)
```
Comment (kiorky) [inline src/croniter/croniter.py:899]: Can you use format expressions ? `"Bad expression: {0}".format(expr)`

File: src/croniter/croniter.py [hunk 4]
```diff
@@ -869,6 +869,9 @@ def expand(self, efl, idx, expr, hash_id=None, match='', **kw):
 
         if m['range_begin'] and m['range_end'] and m['divisor']:
             # Example: H(30-59)/10 -> 34-59/10 (i.e. 34,44,54)
+            if int(m["divisor"]) == 0:
+                raise CroniterBadCronError("Bad expression: %s" % expr)
```
Comment (kiorky) [inline src/croniter/croniter.py:873]: Can you use format expressions ? `"Bad expression: {0}".format(expr)`

File: src/croniter/croniter.py (hunk 2 above)
Comment (kiorky) [reply src/croniter/croniter.py]: way better, will be good after the format expr :)

Comment (cuu508) [reply src/croniter/croniter.py]: Hi @kiorky, I've updated the string formatting bits

**Comment by kiorky** [comment]
Oh sorry for delay, i tthough i had merged this already !!!!

**Pull request merged**